skbio/tree/_tree.py

```diff
@@ -3650,168 +4149,540 @@ def update_result():
 
+            for child in children:
+                del child._start
+                del child._stop
+
+        del self._start
+        del self._stop
+
         return DistanceMatrix(result + result.T, [n.name for n in tip_order])
 
-    def compare_rfd(self, other, proportion=False):
-        """Calculate the Robinson and Foulds symmetric difference.
+    def _compare_topology(
+        self,
+        other,
+        method="subsets",
+        shared_only=True,
+        proportion=False,
+        symmetric=True,
+        include_single=False,
+        weighted=False,
+        metric="euclidean",
+    ):
+        r"""Calculate the topological difference between self and other.
+
+        This function calculates the Robinson-Foulds (RF) distance or its derivates.
 
         Parameters
         ----------
         other : TreeNode
-            A tree to compare against
-        proportion : bool
-            Return a proportional difference
+            The other tree to compare with.
+        method : str, optional
+            Subsets or bipartitions.
+        shared_only : bool, optional
+            Refine to shared taxa.
+        proportion : bool, optional
+            Normalize to fraction.
+        symmetric : bool, optional
+            Symmetric difference.
+        include_single : bool, optional
+            Include singletons.
+        weighted : bool, optional
+            Weight by branch length.
+        metric : str or callable, optional
+            Pairwise distance metric.
 
         Returns
         -------
-        float
-            The distance between the trees
+        int or float
+            Difference between self and other.
+
+        See Also
+        --------
+        compare_rfd
+        compare_wrfd
+        compare_subsets
+        compare_biparts
+
+        """
+        topo1, topo2 = getattr(self, method), getattr(other, method)
+        kwargs = dict(include_single=include_single, map_to_length=weighted)
+        if shared_only:
+            set1, set2 = self.subset(), other.subset()
+            n_shared = len(shared := set1 & set2)
+            sets1 = topo1(within=(shared if len(set1) > n_shared else None), **kwargs)
+            sets2 = topo2(within=(shared if len(set2) > n_shared else None), **kwargs)
+        else:
+            sets1, sets2 = topo1(**kwargs), topo2(**kwargs)
+
+        # unweighted (set difference)
+        if not weighted:
+            method = ("symmetric_" if symmetric else "") + "difference"
+            result = len(getattr(sets1, method)(sets2))
+
+            # normalize result to unit range [0, 1]
+            # if total is 0, return 1 (dist = 1 means saturation)
+            if proportion:
+                total = len(sets1) + (symmetric and len(sets2))
+                result = result / total if total else 1.0
+
+        # branch length weighted (vector distance)
+        else:
+            union = frozenset(sets1.keys()).union(sets2.keys())
```

Comment:
```suggestion
            union = frozenset(sets1).union(sets2)
```

Reply:
I recall that I benchmarked this (either myself or delegated to AI) and found that `keys` is faster than without. But my memory could be incorrect (either I didn't or I found the other way around).

Reply:
Tested. Your suggestion is faster. Will change.

```python
>>> lst = list(range(10000))
>>> dict1 = {x: None for x in random.sample(lst, 5000)}
>>> dict2 = {x: None for x in random.sample(lst, 5000)}
>>> %%timeit
>>> _ = frozenset(dict1).union(dict2)
198 μs ± 6.35 μs per loop
>>> %%timeit
>>> _ = frozenset(dict1.keys()).union(dict2.keys())
259 μs ± 4.02 μs per loop
```

Reply:
Ok, but the suggestion was because the calls to `.keys` here are not necessary, not for the purpose of a micro-optimization (which I recommend against focusing on in this PR).